### PR TITLE
Add union

### PR DESCRIPTION
**Pull request by mitchellwrosen**
I've found this useful a couple times, but I'm not sure if it belongs in this library proper.

### CODE REVIEW

File: src/Numeric/Interval/Internal.hs [hunk 1]
```diff
@@ -673,7 +674,21 @@ intersection x@(I a b) y@(I a' b')
...
+--
+-- >>> union (0 ... 10 :: Interval Double) (5 ... 15 :: Interval Double)
+-- 0.0 ... 15.0
+--
+-- >>> union (15 ... 85 :: Interval Double) (0 ... 10 :: Interval Double)
+-- Empty
+union :: Ord a => Interval a -> Interval a -> Interval a
+union x@(I a b) y@(I a' b')
+  | x /=! y   = Empty
+  | otherwise = I (min a a') (max b b')
```
Comment (dmcclean) [inline src/Numeric/Interval/Internal.hs]:
```
  | otherwise = hull x y
```

Then you could drop the pattern matching of `a`, `b`, `a'`, and `b'`.

Comment (mitchellwrosen) [inline src/Numeric/Interval/Internal.hs]: Hm, ok, updated

**Comment by ekmett** [comment]
This is a bit of a strange operation.

**Comment by ekmett** [comment]
It isn't a hull, and it isn't an intersection, its an 'if and only if they _could_ overlap then extend to the hull', which doesn't sound very union-y to me.

**Comment by mitchellwrosen** [comment]
Indeed. I'm using in a `[Interval a] -> [Interval a]` function that merges a sorted list of intervals (primary lower bound, secondary upper bound), outputting a list of non-overlapping intervals.

That is,

```haskell
merge [1...5, 6...10] = [1...5, 6...10]
merge [1...5, 5...10] = [1...10]
```

Anyways, just thought it might be useful upstream. If not, close away.

**Comment by dmcclean** [comment]
Postgresql's union operator (named `+`) for range types (their name for what we call interval types, their interval type is something else) operates essentially this way. But it throws an exception in the case of non-overlapping ranges.

I'm not sure I get the example use.

Maybe `extend` or `extendIfPossible`, or `combineOverlapping` or something is a better name?

Or maybe we just want the function that takes a set of intervals and shrinks it by outputting non-overlapping intervals with the same coverage, that certainly seems more useful in general and easier to document? Heck, full blown support for `IntervalSet`s would be nice.

**Comment by mitchellwrosen** [comment]
Your last paragraph describes the merge function I was trying to convey =)

**Comment by ekmett** [comment]
A full-on module for `IntervalSet`s would be nice to have and could subsume this usecase.

**Comment by dmcclean** [comment]
See https://hackage.haskell.org/package/IntervalMap for some existing work.

**Comment by ekmett** [comment]
I'm going to go with the close option, just because this has been lingering a long time and the operation still makes me uncomfortable.

**Pull request closed without merging**